doc/watermarks_HLD.md

```diff
@@ -0,0 +1,185 @@
+* [Table 2: Abbreviations](#definitionsabbreviation)
+
+###### Revision
+| Rev |     Date    |       Author       | Change Description                |
+|:---:|:-----------:|:------------------:|-----------------------------------|
+| 0.1 |             | Mykola Faryma      | Initial version                   |
+
+# About this Manual
+This document provides general information about the watermark feature implementation in SONiC.
+# Scope
+This document describes the high level design of the watermark feature.
+# Definitions/Abbreviation
+###### Table 2: Abbreviations
+| Definitions/Abbreviation | Description                                |
+|--------------------------|--------------------------------------------|
+| gRPC                     | gRPC Remote Procedure Calls                |
+| gNMI                     | gRPC Network Management Interface          |
+| API                      | Application Programmable Interface         |
+| SAI                      | Switch Abstraction Interface               |
+
+# 1 Overview
+## 1.1 System Chart
+Following diagram describes a top level overview of the architecture:
+
+![](https://github.com/mykolaf/SONiC/blob/gh-pages/images/watermark_HLD/SystemOverview.png)
+
+## 1.2 Modules description
+### 1.2.1 gRPC
+System data telemetry infrastructure. Basically allows to getRequest data from SONiC DBs (and more).
+### 1.2.2 Counter DB
+Located in the Redis DB instance #2 running inside the container "database". Redis DB works with the data in format of key-value tuples, needs no predefined schema and holds various counters like port counters, ACL counters, etc.
+### 1.2.3 Orchestration Agent
+This component is running in the "orchagent" docker container and is responsible for processing updates of the APP DB and do corresponding changes in the SAI DB via SAI Redis.
+### 1.2.4 SAI Redis
+SAI Redis is an implementation of the SAI API which translates API calls into SAI objects which are stored in the ASIC DB.
+### 1.2.5 ASIC DB
+Redis DB instance #1. Holds serialized SAI objects.
+### 1.2.6 syncd
+Reads SAI DB data (SAI objects) and performs appropriate calls to Switch SAI.
+### 1.2.7 SAI (Redis and Switch)
+An unified API which represent the switch state as a set of objects. In SONiC represented in two implementations - SAI DB frontend and ASIC SDK wrapper.
+# 2 Requirements
+The following watermarks should be supported:
+## 2.1 Watermark counters requirements
+|                                          | SAI attribute mapping  |
+|------------------------------------------|------------------------|
+| Ingress headroom per PG                  | SAI_INGRESS_PRIORITY_GROUP_STAT_XOFF_ROOM_WATERMARK_BYTES            |
+| Ingress shared pool occupancy per PG                | SAI_INGRESS_PRIORITY_GROUP_STAT_SHARED_WATERMARK_BYTES            |
+| Egress shared pool occupancy per queue (including both unicast queues and multicast queues)      | SAI_QUEUE_STAT_SHARED_WATERMARK_BYTES         |
```

Comment:
do we have the total pool watermark here? I think we need that as well.

Reply:
Pool watermark was not in the requirements. Also as far as I know pool objects are not created in SONiC, so more complexity.